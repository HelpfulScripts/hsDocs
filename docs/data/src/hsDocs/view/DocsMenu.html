
                <html><style>body { overflow:hidden;}
            h1 { font-family: Arial, sans-serif; font-size: 24px; color: #44a; }
            p { margin:0; padding:0; }
            br  { margin:0; padding:0; }
            .line { 
                white-space: nowrap;
                height:16px; 
            }
            .line>span { 
                display:inline-block; background-color:#eef; height:100%; 
                margin: 0 5px 0 0; padding-right: 5px; color:#999;   
            }
            comment {color: #080;} module {color: #804;} 
            quote {color: #008;} comment>quote {color: #080;}
            .listing { margin: 10px; border: 1px solid #ccc; 
                    font-family: SFMono-Regular, Consolas, 'Liberation Mono', Menlo, Courier, monospace;
                    font-size: 14px;  
                    overflow:scroll;
                    height:90%;
            }
            code { padding: 5px 0;}</style></html><body>
                <h1>src/view/DocsMenu.ts</h1>
                <div class='listing'><code><p id=1 class="line"><span>&nbsp;&nbsp;&nbsp;1</span>import&nbsp;{&nbsp;m,&nbsp;Vnode,&nbsp;Layout&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hslayout'</quote>;</p>
<p id=2 class="line"><span>&nbsp;&nbsp;&nbsp;2</span>import&nbsp;{&nbsp;DocSets&nbsp;}&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'../DocSets'</quote>;&nbsp;</p>
<p id=3 class="line"><span>&nbsp;&nbsp;&nbsp;3</span>import&nbsp;{&nbsp;Menu,&nbsp;SelectorDesc&nbsp;}&nbsp;&nbsp;&nbsp;from&nbsp;<quote>'hswidget'</quote>;</p>
<p id=4 class="line"><span>&nbsp;&nbsp;&nbsp;4</span>&nbsp;</p>
<p id=5 class="line"><span>&nbsp;&nbsp;&nbsp;5</span><comment>/**</comment></p>
<p id=6 class="line"><span>&nbsp;&nbsp;&nbsp;6</span><comment>&nbsp;*&nbsp;Creates&nbsp;the&nbsp;title&nbsp;menu&nbsp;for&nbsp;selecting&nbsp;between&nbsp;the&nbsp;different&nbsp;docsets.</comment></p>
<p id=7 class="line"><span>&nbsp;&nbsp;&nbsp;7</span><comment>&nbsp;*&nbsp;Instantiate&nbsp;the&nbsp;DocsMenu&nbsp;via&nbsp;a&nbsp;standard&nbsp;`mithril`&nbsp;call:```</comment></p>
<p id=8 class="line"><span>&nbsp;&nbsp;&nbsp;8</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;&nbsp;m(DocsMenu,&nbsp;{&nbsp;docSet:<pathToIndexFile>})</comment></p>
<p id=9 class="line"><span>&nbsp;&nbsp;&nbsp;9</span><comment>&nbsp;*&nbsp;```</comment></p>
<p id=10 class="line"><span>&nbsp;&nbsp;10</span><comment>&nbsp;*&nbsp;DocsMenu&nbsp;performs&nbsp;the&nbsp;following&nbsp;actions:</comment></p>
<p id=11 class="line"><span>&nbsp;&nbsp;11</span><comment>&nbsp;*&nbsp;-&nbsp;for&nbsp;the&nbsp;first&nbsp;call&nbsp;of&nbsp;the&nbsp;view&nbsp;lifecycle&nbsp;hook,&nbsp;the&nbsp;available&nbsp;docSets&nbsp;are&nbsp;loaded.</comment></p>
<p id=12 class="line"><span>&nbsp;&nbsp;12</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;DocsMenu&nbsp;searches&nbsp;for&nbsp;an&nbsp;index&nbsp;`json`&nbsp;file&nbsp;at&nbsp;the&nbsp;location&nbsp;specified&nbsp;in&nbsp;the&nbsp;</comment></p>
<p id=13 class="line"><span>&nbsp;&nbsp;13</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;`docSet`&nbsp;field&nbsp;of&nbsp;the&nbsp;`node.attrs`&nbsp;parameter.&nbsp;If&nbsp;none&nbsp;is&nbsp;specified,&nbsp;the&nbsp;</comment></p>
<p id=14 class="line"><span>&nbsp;&nbsp;14</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;default&nbsp;is&nbsp;used&nbsp;as&nbsp;specified&nbsp;in&nbsp;the&nbsp;{@link&nbsp;hsDocs:DocSets.FILE&nbsp;DocSets&nbsp;FILE}&nbsp;setting.</comment></p>
<p id=15 class="line"><span>&nbsp;&nbsp;15</span><comment>&nbsp;*&nbsp;-&nbsp;DocsMenu&nbsp;retrieves&nbsp;all&nbsp;available&nbsp;docSets&nbsp;via&nbsp;{@link&nbsp;hsDocs:DocSets.DocSets.get&nbsp;DocSets.get}.</comment></p>
<p id=16 class="line"><span>&nbsp;&nbsp;16</span><comment>&nbsp;*&nbsp;-&nbsp;DocsMenu&nbsp;creates&nbsp;a&nbsp;`SelectorDesc`&nbsp;structure&nbsp;with&nbsp;a&nbsp;{@link&nbsp;hsWidget:hsSelector.SelectorDesc.changed&nbsp;`changed`}&nbsp;callback&nbsp;that&nbsp;initiates&nbsp;a&nbsp;route&nbsp;change&nbsp;</comment></p>
<p id=17 class="line"><span>&nbsp;&nbsp;17</span><comment>&nbsp;*&nbsp;&nbsp;&nbsp;to&nbsp;the&nbsp;selected&nbsp;docSet</comment></p>
<p id=18 class="line"><span>&nbsp;&nbsp;18</span><comment>&nbsp;*/</comment></p>
<p id=19 class="line"><span>&nbsp;&nbsp;19</span>export&nbsp;class&nbsp;DocsMenu&nbsp;extends&nbsp;Layout&nbsp;{</p>
<p id=20 class="line"><span>&nbsp;&nbsp;20</span>&nbsp;&nbsp;&nbsp;&nbsp;docSet&nbsp;=&nbsp;<quote>''</quote>;</p>
<p id=21 class="line"><span>&nbsp;&nbsp;21</span></p>
<p id=22 class="line"><span>&nbsp;&nbsp;22</span>&nbsp;&nbsp;&nbsp;&nbsp;private&nbsp;getDesc(attrs:any):SelectorDesc&nbsp;{&nbsp;</p>
<p id=23 class="line"><span>&nbsp;&nbsp;23</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;if&nbsp;(this.docSet&nbsp;!==&nbsp;attrs.docSet)&nbsp;{</p>
<p id=24 class="line"><span>&nbsp;&nbsp;24</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;this.docSet&nbsp;=&nbsp;attrs.docSet;</p>
<p id=25 class="line"><span>&nbsp;&nbsp;25</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;DocSets.loadList(attrs.docSet);</p>
<p id=26 class="line"><span>&nbsp;&nbsp;26</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=27 class="line"><span>&nbsp;&nbsp;27</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;items&nbsp;=&nbsp;DocSets.get();&nbsp;</p>
<p id=28 class="line"><span>&nbsp;&nbsp;28</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;{</p>
<p id=29 class="line"><span>&nbsp;&nbsp;29</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;items:&nbsp;items.map((c:string)&nbsp;=>&nbsp;c),</p>
<p id=30 class="line"><span>&nbsp;&nbsp;30</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;defaultItem:&nbsp;(attrs.route&nbsp;&&&nbsp;attrs.route.lib)?&nbsp;attrs.route.lib&nbsp;:&nbsp;items[0],</p>
<p id=31 class="line"><span>&nbsp;&nbsp;31</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;changed:&nbsp;(item:string)&nbsp;=>&nbsp;m.route.set(<quote>'/api/:lib/0'</quote>,&nbsp;{lib:item})</p>
<p id=32 class="line"><span>&nbsp;&nbsp;32</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;};</p>
<p id=33 class="line"><span>&nbsp;&nbsp;33</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=34 class="line"><span>&nbsp;&nbsp;34</span></p>
<p id=35 class="line"><span>&nbsp;&nbsp;35</span>&nbsp;&nbsp;&nbsp;&nbsp;getComponents(node:Vnode):Vnode&nbsp;{</p>
<p id=36 class="line"><span>&nbsp;&nbsp;36</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;const&nbsp;desc:SelectorDesc&nbsp;=&nbsp;this.getDesc(node.attrs);</p>
<p id=37 class="line"><span>&nbsp;&nbsp;37</span>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return&nbsp;m(Menu,&nbsp;{desc:&nbsp;desc});</p>
<p id=38 class="line"><span>&nbsp;&nbsp;38</span>&nbsp;&nbsp;&nbsp;&nbsp;}</p>
<p id=39 class="line"><span>&nbsp;&nbsp;39</span>}</p></code></div>
                </body>
            