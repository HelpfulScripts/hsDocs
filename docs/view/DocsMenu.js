"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const hslayout_1 = require("hslayout");
const DocSets_1 = require("../DocSets");
const hswidget_1 = require("hswidget");
class DocsMenu extends hslayout_1.Layout {
    constructor() {
        super(...arguments);
        this.docSet = '';
    }
    getDesc(attrs) {
        if (this.docSet !== attrs.docSet) {
            this.docSet = attrs.docSet;
            DocSets_1.DocSets.loadList(attrs.docSet);
        }
        const items = DocSets_1.DocSets.get();
        return {
            items: items.map((c) => c),
            defaultItem: (attrs.route && attrs.route.lib) ? attrs.route.lib : items[0],
            changed: (item) => hslayout_1.m.route.set('/api/:lib/0', { lib: item })
        };
    }
    getComponents(node) {
        const desc = this.getDesc(node.attrs);
        return hslayout_1.m(hswidget_1.Menu, { desc: desc });
    }
}
exports.DocsMenu = DocsMenu;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoiRG9jc01lbnUuanMiLCJzb3VyY2VSb290IjoiIiwic291cmNlcyI6WyIuLi8uLi9zcmMvdmlldy9Eb2NzTWVudS50cyJdLCJuYW1lcyI6W10sIm1hcHBpbmdzIjoiOztBQUFBLHVDQUFnRDtBQUNoRCx3Q0FBa0Q7QUFDbEQsdUNBQWdEO0FBZ0JoRCxjQUFzQixTQUFRLGlCQUFNO0lBQXBDOztRQUNJLFdBQU0sR0FBRyxFQUFFLENBQUM7SUFtQmhCLENBQUM7SUFqQlcsT0FBTyxDQUFDLEtBQVM7UUFDckIsSUFBSSxJQUFJLENBQUMsTUFBTSxLQUFLLEtBQUssQ0FBQyxNQUFNLEVBQUU7WUFDOUIsSUFBSSxDQUFDLE1BQU0sR0FBRyxLQUFLLENBQUMsTUFBTSxDQUFDO1lBQzNCLGlCQUFPLENBQUMsUUFBUSxDQUFDLEtBQUssQ0FBQyxNQUFNLENBQUMsQ0FBQztTQUNsQztRQUNELE1BQU0sS0FBSyxHQUFHLGlCQUFPLENBQUMsR0FBRyxFQUFFLENBQUM7UUFDNUIsT0FBTztZQUNILEtBQUssRUFBRSxLQUFLLENBQUMsR0FBRyxDQUFDLENBQUMsQ0FBUSxFQUFFLEVBQUUsQ0FBQyxDQUFDLENBQUM7WUFDakMsV0FBVyxFQUFFLENBQUMsS0FBSyxDQUFDLEtBQUssSUFBSSxLQUFLLENBQUMsS0FBSyxDQUFDLEdBQUcsQ0FBQyxDQUFBLENBQUMsQ0FBQyxLQUFLLENBQUMsS0FBSyxDQUFDLEdBQUcsQ0FBQyxDQUFDLENBQUMsS0FBSyxDQUFDLENBQUMsQ0FBQztZQUN6RSxPQUFPLEVBQUUsQ0FBQyxJQUFXLEVBQUUsRUFBRSxDQUFDLFlBQUMsQ0FBQyxLQUFLLENBQUMsR0FBRyxDQUFDLGFBQWEsRUFBRSxFQUFDLEdBQUcsRUFBQyxJQUFJLEVBQUMsQ0FBQztTQUNuRSxDQUFDO0lBQ04sQ0FBQztJQUVELGFBQWEsQ0FBQyxJQUFVO1FBQ3BCLE1BQU0sSUFBSSxHQUFnQixJQUFJLENBQUMsT0FBTyxDQUFDLElBQUksQ0FBQyxLQUFLLENBQUMsQ0FBQztRQUNuRCxPQUFPLFlBQUMsQ0FBQyxlQUFJLEVBQUUsRUFBQyxJQUFJLEVBQUUsSUFBSSxFQUFDLENBQUMsQ0FBQztJQUNqQyxDQUFDO0NBQ0o7QUFwQkQsNEJBb0JDIn0=