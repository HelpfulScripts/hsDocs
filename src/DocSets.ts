/**
 * DocSets.ts. Loads the doc.json files to process and display as documentation.
 * Processing occurs in these steps:
 * 1. Load the index.json file that contains a list of doc.json files to load, one for each library to show
 * 2. Load each doc.json file, which describes a library
 * 3. Call DocSets.add to add the library name to the registry and create an index of entries for the library
 */

/** */
import { m }            from 'hslayout';
import { Log }          from 'hsutil'; const log = new Log('DocsSets');
import { DocsNode }     from './Nodes';

const DOCDIR:string = './data/';

/**
 * the default location for the index .json files, relative to the web page:
 * `'./data/index.json'`
 */
const FILE:string = './data/index.json';

export type json = any;

function matchAll(str:string, re:RegExp): string[] {
    const result:string[] = [];
    let partial:string[];
    while ((partial = re.exec(str)) !== null) {
        result.push(partial[1]);
        log.debug(log.inspect(partial));
    }
    return result;
}

/**
 * DocSets object. Keeps a list of registered docsets and 
 * provides access to elements of each docset.
 */
export class DocSets { 
    /** Contains references to the docsets and all elements per docset, accessible per ID. */
    // private static gList:List = {index:{}};

    private static libs:string[] = [];
    private static docs:string[] = [];

    /** an optional title for the set of DocSets loaded, as specified in the `index.json` file. */
    private static gTitle: string = '';
    private static nodeList = <{[id:string]: DocsNode}>{};
    
    public static nodeCount = 0;
    
    public static addNode(mdl:json, node:DocsNode) {
        const id = mdl.id;
        const lib =  mdl.lib;
        DocSets.nodeList[`${lib}.${id}`] = node;
        DocSets.nodeList[mdl.fullPath] = node;
// if (mdl.fullPath.indexOf('hsDocs.DocSets')>=0) { log.info(`added node ${id}: '${mdl.fullPath}'`); }
        DocSets.nodeCount++;
    }


    /**
     * Adds the docset in `content` to the `gList` at the position of `file` in `DocSets.docs`.
     * @param content the docSet content to add. 
     * @param file the file name from which the content was read. 
     */
    public static addDocSet(content:json, file:string):void {
        const lib = content.name;
        const i = DocSets.docs.indexOf(file);
        DocSets.libs[i] = lib;
        const root = DocsNode.traverse(content, lib);
        log.info(`traversed '${root.fullPath}'`);
    }
    
    public static getNode(id:string|number, lib:string):DocsNode {
        const key = ((typeof id === 'number') || (parseInt(''+id, 10)>=0))? `${lib}.${id}` : id;
        // log.info(`getNode id=${id}, lib=${lib} -> ${key}, typeof=${typeof id}, parseInt=${parseInt(''+id, 10)}`);
        if (DocSets.nodeList.length && !DocSets.nodeList[key]) { 
            log.warn(`did not find node for key '${key}' (id=${id}, lib=${lib})`); 
            log.warn(new Error().stack);
        }
        return DocSets.nodeList[key]; 
    }
    
    /**
     * returns the specified documentation element.
     * When called without parameters, a `string[lib]` of available docSets is returned.
     * When called with only `lib` specified, the corresponding docSet overview is returned.
     * @param lib specifies the docset to scan. 
     * @param id specifies the element within the docSet, either by its id number, or by its fully qualified path, 
     * e.g. 'hsDocs.DocSets.DocSets.add' 
     */
    public static getLibs():string[]{ 
        return DocSets.libs; 
    }

    /**
     * loads an index set and the docsets it contains from directory `dir`.
     * @param file the optional directory to read from. If unspecified, 
     * read from the index file specified by {@link DocSets.FILE `FILE`}.
     * @return a promise to load the index set
     */
    public static async loadList(file?:string):Promise<void> {
        /** load an `index.json` file that contains references to the DocSets to load. */
        async function getIndexFile(dir:string, url:string):Promise<boolean> {
            if (!url) { return false; }
            try {
                const result = await m.request({ method: "GET", url: file});
                DocSets.gTitle = result.title;
                DocSets.docs = result.docs.map((doc:string) => doc.indexOf(':')>0? doc : dir+doc);
                log.info(`found index file ${url} with ${DocSets.docs.length} library references`);
                return true;
            } catch(e) {
                return false;
            }
        }
        /** 
         * get all `jsons` in a directory specified by `url`. 
         * If the directory contains an `index.json`, calls `getIndexFile` on that file's url.
         */
        async function getDirJSONs(url:string):Promise<boolean> {
            let result = false;
            await m.request({ method: "GET", url: url, 
                // catch extract to avoid attempt to deserialize
                extract: async (xhr:any, options:any) => { 
                    const jsons = matchAll(xhr.responseText, /<a href="(\S*?.json)">/g);  // all *.json references
                    // check if index.jon exists:
                    const index = jsons.filter(m => m==='index.json')[0];
                    // const jsons = matches.filter(m => !m.startsWith('index'));

                    if (index) { 
                        result = await getIndexFile(dir, index); 
                    } else if (jsons.length) {
                        log.info(`found dir list in ${url}`);
                        DocSets.gTitle = '';
                        DocSets.docs = jsons;
                        result = true;
                    }
                }
            });
            return result;
        }
        let i = file.lastIndexOf('/');
        const dir = file.substring(0,i+1);
        
        let found = false;
        if (!found) { found = await getIndexFile(dir, file); }
        if (!found) { found = await getDirJSONs(dir); }
        if (!found) { found = await getDirJSONs(DOCDIR); }
        log.debug(`found ${DocSets.docs.length} dos sets: ${log.inspect(DocSets.docs, 5)}`);
        await Promise.all(DocSets.docs.map(async (f:string) => await loadDocSet(f))).catch(log.error);        
        log.info(`found ${DocSets.nodeCount} DocNodes`);
        m.redraw();
    }
}

/**
 * Loads a docset specified by file from the directory `dir`. 
 * Once received, the docset is registered in `modules` via the `add` method.
 * @param dir the directory to read from
 * @param file the `json` file to load as docset
 */
async function loadDocSet(file:string):Promise<void> {
    log.debug(`loading ${file}`);
    const r:json = await m.request({ method: "GET", url: file });
    DocSets.addDocSet(r, file);
}

