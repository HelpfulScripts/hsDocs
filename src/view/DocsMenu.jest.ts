import { DocsMenu } from './DocsMenu';
import { DocSets }  from '../DocSets'; 
import m from "mithril";

const sets = {
    "docs": [
        "hsUtil.json",
        "hsDocs.json"
    ],
    "title": "HS Libraries"
};

const set1 = {
	"id": 0,
	"name": "hsUtil",
	"kind": 0,
	"flags": {},
	"children": [{
        "id": 1,
        "name": "\"Checksum\"",
        "kind": 1,
        "kindString": "External module",
        "flags": {},
        "originalName": "/Users/sth1pal/Documents/Development/JavaScript/nodejs/ts6/dev/hsLibs/standalone/hsUtil/src/Checksum.ts",
        "comment": {},
        "groups": <any[]>[],
        "sources": <any[]>[]
    }]
};

const set2 = {
	"id": 0,
	"name": "hsDocs",
	"kind": 0,
	"flags": {},
	"children": [{
        "id": 1,
        "name": "\"Checksum\"",
        "kind": 1,
        "kindString": "External module",
        "flags": {},
        "originalName": "/Users/sth1pal/Documents/Development/JavaScript/nodejs/ts6/dev/hsLibs/standalone/hsUtil/src/Checksum.ts",
        "comment": <any>{},
        "groups": <any[]>[],
        "sources": <any[]>[]
    }]
};

m.request = (req: any) => {
    let result;
    if (req.url === 'mylist.json') {
        result = sets;
    } else if (req.url === 'hsUtil.json'){
        result = set1;
    } else {
        result = set2;
    }
    return Promise.resolve(result).then(r => { m.redraw(); return r; });
};


const root:any = window.document.createElement("div");

describe('DocsMenu', () => {
    beforeAll(async (done) => { 
        await DocSets.loadList("mylist.json");
        m.mount(root, { view: () => m(DocsMenu, {docSet:"mylist.json"}) }); 
        setTimeout(done, 200);
    });

    it('should match snapshot', (done) => {
        expect(root).toMatchSnapshot();
        done();
    });

    // it ('should respond to click', (done) => {
    //     const menuItem = root.vnodes[0].instance.instance.children[0].instance.children[0].instance.children[1].instance.children[0];
    //     menuItem.events.onclick('hsDocs');
    //    setTimeout(()=>{
    //        console.log(menuItem);
    //    }, 100); 
    // });
});
