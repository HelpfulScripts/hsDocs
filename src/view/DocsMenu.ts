import { m, Vnode, Layout }         from 'hslayout';
import { DocSets }                  from '../DocSets'; 
import { Menu, SelectorDesc }       from 'hswidget';
import { log as gLog }              from 'hsutil'; const log = gLog('DocsMenu');

/**
 * Creates the title menu for selecting between the different docsets.
 * Instantiate the DocsMenu via a standard `mithril` call:```
 *    m(DocsMenu, { docSet:<pathToIndexFile>})
 * ```
 * DocsMenu performs the following actions:
 * - for the first call of the view lifecycle hook, the available docSets are loaded.
 *   DocsMenu searches for an index `json` file at the location specified in the 
 *   `docSet` field of the `node.attrs` parameter. If none is specified, the 
 *   default is used as specified in the {@link hsDocs:DocSets.FILE DocSets FILE} setting.
 * - DocsMenu retrieves all available docSets via {@link hsDocs:DocSets.DocSets.get DocSets.get}.
 * - DocsMenu creates a `SelectorDesc` structure with a {@link hsWidget:hsSelector.SelectorDesc.clicked `clicked`} callback that initiates a route change 
 *   to the selected docSet
 */
export class DocsMenu extends Layout {
    docSet = '';

    private getDesc(attrs:any):SelectorDesc { 
        if (this.docSet !== attrs.docSet) {
            this.docSet = attrs.docSet;
            DocSets.loadList(attrs.docSet)
            .then(() => {
                m.redraw();
                DocSets.getLibs().map((set:string) => {
                    const mdl = DocSets.getNode(0, set);
                });
            })
            .catch(e => {
                log.error(e.message);
                log.error(e);
            });
        }
        const items = DocSets.getLibs(); 
        return {
            items: items.map((c:string) => c),
            // defaultItem: (attrs.route && attrs.route.lib)? attrs.route.lib : items[0],
            defaultItem: m.route.param('lib') || items[0],
            clicked: (item:string) => m.route.set('/api/:lib/0', {lib:item})
        };
    }

    getComponents(node:Vnode):Vnode {
        const desc:SelectorDesc = this.getDesc(node.attrs);
        return m(Menu, {desc: desc});
    }
}