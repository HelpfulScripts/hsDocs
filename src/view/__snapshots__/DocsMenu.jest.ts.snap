// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocsMenu should match snapshot 1`] = `
<div>
  <div
    class="hs_option_buttons  hs_radio_buttons hsdocs_menu hs_menu "
    style="grid-template-columns: repeat(2, 1fr);"
  >
    <div
      class="hs_button state0  hs_onoff_button index0 "
    >
      hsUtil
    </div>
    <div
      class="hs_button state0  hs_onoff_button index1 "
    >
      hsDocs
    </div>
  </div>
</div>
`;
